gpytorch/priors/prior.py

```diff
@@ -0,0 +1,66 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+from abc import abstractmethod, abstractproperty
+import torch
+from torch.nn import Module
+
+
+class Prior(Module):
+    """
+    Base class for Priors in GPyTorch.
+    In GPyTorch, a parameter can be assigned a prior by passing it as the `prior` argument to
+    :func:`~gpytorch.module.register_parameter`. GPyTorch performs internal bookkeeping of priors,
+    and for each parameter with a registered prior includes the log probability of the parameter under its
+    respective prior in computing the Marginal Log-Likelihood (see e.g. :func:`~gpytorch.priors.Prior.forward`).
+
+    In order to define a new Prior in GPyTorch, a user must define at a minimum the following methods
+    * :func:`~gpytorch.priors.Prior.shape`, which returns the shape of the domain as a torch.Size.
+    * :func:`~gpytorch.priors.Prior.is_in_support`, which for a given parameter value returns a bool indicating
+        whether the parameter is contained in the support of the distribution.
+    * :func:`~gpytorch.priors.Prior._log_prob`, which returns the log-probability for a given parameter value
+        as a torch scalar.
+
+    """
+
+    @abstractmethod
+    def is_in_support(self, parameter):
+        raise NotImplementedError()
+
+    @property
+    def log_transform(self):
+        return self._log_transform
+
+    @abstractmethod
+    def _log_prob(self, parameter):
+        raise NotImplementedError()
+
+    @abstractproperty
+    def shape(self):
+        raise NotImplementedError()
+
+    def log_prob(self, parameter):
+        """Returns the log-probability of the parameter value under the prior."""
+        return self._log_prob(parameter.exp() if self.log_transform else parameter)
+
+    def _initialize_distributions(self):
```

Comment:
What is this method supposed to do?

Reply:
Oh I see... it's necessary for the apply function. maybe Priors should have a default constructor that calls this method?

Reply:
sure, I can add this

Reply:
Actually, let's not do this. We need the Module init call before we can assign buffers, but some of the `_initialize_distributions` calls will require some buffers to be set. So it'd be quite awkward doing this if the base class constructor called `_initialize_distributions`.